File: r/tests/test_basic.R
#!/usr/bin/env Rscript
# Basic tests for fast-hdfe-reg package

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/polars.R")
source("R/duckdb.R")
source("R/leanfe.R")

set.seed(42)
n <- 1000

cat("Testing basic functionality...\n\n")

# Test 1: Basic Polars functionality
cat("Test 1: Basic Polars functionality... ")
df <- pl$DataFrame(
  y = rnorm(n),
  x = rnorm(n),
  fe1 = sample(0:9, n, replace = TRUE),
  fe2 = sample(0:4, n, replace = TRUE)
)

result <- leanfe_polars(df, y_col = "y", x_cols = "x", fe_cols = c("fe1", "fe2"))

stopifnot("coefficients" %in% names(result))
stopifnot("std_errors" %in% names(result))
stopifnot("x" %in% names(result$coefficients))
stopifnot(result$n_obs > 0)
cat("PASSED\n")

# Test 2: Formula API
cat("Test 2: Formula API... ")
result <- leanfe_polars(df, formula = "y ~ x | fe1 + fe2")

stopifnot("coefficients" %in% names(result))
stopifnot("x" %in% names(result$coefficients))
cat("PASSED\n")

# Test 3: Standard errors (IID, HC1, Clustered)
cat("Test 3: Standard errors... ")
df_cluster <- pl$DataFrame(
  y = rnorm(n),
  x = rnorm(n),
  fe1 = sample(0:9, n, replace = TRUE),
  cluster = sample(0:19, n, replace = TRUE)
)

# IID
result_iid <- leanfe_polars(df_cluster, y_col = "y", x_cols = "x", fe_cols = "fe1", vcov = "iid")
stopifnot(result_iid$vcov_type == "iid")

# HC1
result_hc1 <- leanfe_polars(df_cluster, y_col = "y", x_cols = "x", fe_cols = "fe1", vcov = "HC1")
stopifnot(result_hc1$vcov_type == "HC1")

# Clustered
result_cluster <- leanfe_polars(
  df_cluster, 
  y_col = "y", 
  x_cols = "x", 
  fe_cols = "fe1",
  vcov = "cluster", 
  cluster_cols = "cluster"
)
stopifnot(result_cluster$vcov_type == "cluster")
stopifnot(result_cluster$n_clusters > 0)
cat("PASSED\n")

# Test 4: Unified API with Polars backend
cat("Test 4: Unified API (backend='polars')... ")
result_unified <- leanfe(df, formula = "y ~ x | fe1 + fe2", backend = "polars")
stopifnot("x" %in% names(result_unified$coefficients))
cat("PASSED\n")

# Test 5: Unified API with DuckDB backend
cat("Test 5: Unified API (backend='duckdb')... ")
result_duckdb <- leanfe(df, formula = "y ~ x | fe1 + fe2", backend = "duckdb")
stopifnot("x" %in% names(result_duckdb$coefficients))
cat("PASSED\n")

# Test 6: Unified API default backend (polars)
cat("Test 6: Unified API (default backend)... ")
result_default <- leanfe(df, formula = "y ~ x | fe1 + fe2")
stopifnot("x" %in% names(result_default$coefficients))
cat("PASSED\n")

# Test 7: Unified API invalid backend raises error
cat("Test 7: Unified API (invalid backend)... ")
error_caught <- FALSE
tryCatch({
  leanfe(df, formula = "y ~ x | fe1", backend = "invalid")
}, error = function(e) {
  if (grepl("backend must be", e$message)) {
    error_caught <<- TRUE
  }
})
stopifnot(error_caught)
cat("PASSED\n")

cat("\n✅ All 7 tests passed!\n")

File: python/tests/test_basic.py
"""Basic tests for fast-hdfe-reg package."""

import polars as pl
import numpy as np
from leanfe import leanfe, leanfe_polars, leanfe_duckdb


def test_polars_basic():
    """Test basic Polars functionality."""
    # Create simple test data
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "fe2": np.random.randint(0, 5, n),
    })
    
    result = leanfe_polars(df, "y", ["x"], ["fe1", "fe2"])
    
    assert "coefficients" in result
    assert "std_errors" in result
    assert "x" in result["coefficients"]
    assert result["n_obs"] > 0


def test_formula_api():
    """Test formula parsing API."""
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "fe2": np.random.randint(0, 5, n),
    })
    
    result = leanfe_polars(df, formula="y ~ x | fe1 + fe2")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]


def test_standard_errors():
    """Test different standard error types."""
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "cluster": np.random.randint(0, 20, n),
    })
    
    # IID
    result_iid = leanfe_polars(df, "y", ["x"], ["fe1"], vcov="iid")
    assert result_iid["vcov_type"] == "iid"
    
    # HC1
    result_hc1 = leanfe_polars(df, "y", ["x"], ["fe1"], vcov="HC1")
    assert result_hc1["vcov_type"] == "HC1"
    
    # Clustered
    result_cluster = leanfe_polars(
        df, "y", ["x"], ["fe1"], 
        vcov="cluster", cluster_cols=["cluster"]
    )
    assert result_cluster["vcov_type"] == "cluster"
    assert result_cluster["n_clusters"] > 0


def test_unified_api_polars():
    """Test unified leanfe() with Polars backend."""
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "fe2": np.random.randint(0, 5, n),
    })
    
    result = leanfe(df, formula="y ~ x | fe1 + fe2", backend="polars")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]


def test_unified_api_duckdb():
    """Test unified leanfe() with DuckDB backend."""
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "fe2": np.random.randint(0, 5, n),
    })
    
    result = leanfe(df, formula="y ~ x | fe1 + fe2", backend="duckdb")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]


def test_unified_api_default_backend():
    """Test unified leanfe() uses Polars by default."""
    np.random.seed(42)
    n = 1000
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "fe2": np.random.randint(0, 5, n),
    })
    
    # No backend specified - should use polars
    result = leanfe(df, formula="y ~ x | fe1 + fe2")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]


def test_unified_api_invalid_backend():
    """Test unified leanfe() raises error for invalid backend."""
    np.random.seed(42)
    n = 100
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
    })
    
    import pytest
    with pytest.raises(ValueError, match="backend must be"):
        leanfe(df, formula="y ~ x | fe1", backend="invalid")
